src/citrine/builders/design_spaces.py

```diff
@@ -0,0 +1,205 @@
+        {'component name':[list, of, values]} for each component
+    balance_component: str
+        name of the component that should make up the balance of the
+        mixture (1-others). Must be a key in formulation_grids.
+    descriptors: List[Descriptor]
+        design descriptors representing the degrees of freedom of the design
+        space. descriptors' keys should match the keys of formulation_grids.
+        If none are passed, they will be auto-generated based on the keys
+        of formulation_grids.
+    name: str
+        name for the EnumeratedDesignSpace
+    description: str
+        description for the EnumeratedDesignSpace
+
+    """
+    # Generate descriptors if none passed
+    if descriptors is None:
+        descriptors = [RealDescriptor(kk, 0, 1) for kk in formulation_grids.keys()]
+
+    # Check that the passed balance component is in the grid keys
+    if balance_component not in list(formulation_grids.keys()):
+        raise ValueError("balance_component must be in formulation_grids' keys")
+
+    # Check that all grid values are between 0 and 1
+    for kk, val_list in formulation_grids.items():
+        if not (all([x >= 0 for x in val_list]) and all([x <= 1 for x in val_list])):
+            raise ValueError('values in {} are not between 0 and 1'.format(kk))
+
+    non_balance_comps = [comp for comp in list(formulation_grids.keys())
+                         if comp != balance_component]
+    non_balance_descriptors = [dd for dd in descriptors
+                               if dd.key != balance_component]
+    non_balance_grids = {
+        k: v for k, v in formulation_grids.items()
+        if k in non_balance_comps
+    }
+
+    # Start by making a naive product design space of non-balance components
+    form_ds = pd.DataFrame(
+        enumerate_cartesian_product(
+            design_grids=non_balance_grids,
+            descriptors=non_balance_descriptors
+        ).data
+    )
+
+    # Re-calculate the balance component column as 1-all others
+    form_ds[balance_component] = form_ds.apply(
+        lambda r: 1 - sum([r[col] for col in non_balance_comps]),
+        axis=1
+    )
+
+    # Order columns so balance component is first
+    form_ds = form_ds[[balance_component] + non_balance_comps]
+
+    # Eliminate out-of-range rows
+    balance_range = (
+        min(formulation_grids[balance_component]),
+        max(formulation_grids[balance_component])
+    )
+    form_ds = form_ds[form_ds[balance_component] >= balance_range[0]]
+    form_ds = form_ds[form_ds[balance_component] <= balance_range[1]]
+    data = form_ds.to_dict('records')
+    design_space = EnumeratedDesignSpace(name=name,
+                                         description=description,
+                                         descriptors=descriptors,
+                                         data=data)
+    return design_space
+
+
+def cartesian_join_design_spaces(
+    design_space_list: List[EnumeratedDesignSpace],
```

Comment:
`design_spaces`? or `subspaces`?  We probably want to use the same language as https://github.com/CitrineInformatics/ai-engine-documentation/blob/master/docs/design-spaces.md#productdesignspace-modifications-to